### PR TITLE
Fix selections on plots with multiple glyph renderers

### LINKED ISSUE #2136: Selections only work on the last of two series in one plot
If a plot contains multiple series of data, only selections (`box_select` or `tap`) on the second have an effect on the plot, selection data points of the first series does nothing. Both series share the same data source.
### Selection on the second series (green)

screenshot
### Selection on the first series (blue)

screenshot
### Code

Here is the example code:

```
import bokeh.plotting as bk
bk.output_file("example.html",)
chartData = {"name": [1, 2, 3],
                     "number": [0.5, 1, 1.5],
                     "number2": [1.5, 2, 2.5]}
df = bk.ColumnDataSource(data = chartData)
p = bk.figure(title = "selection", tools = "tap,box_select")
p.circle(x = "name", y = "number", source = df)
p.circle(x = "name", y = "number2", color = "green", source = df)
bk.show(p)
```

Bokeh 0.8.2
Python 3.4.3

### LINKED ISSUE #5044: Shift selection in linked brushing plots
I expected to have shift selection in between two linked plots to select the union in the shared common column data source. For instance in the linked brushing example in the docs  http://bokeh.pydata.org/en/latest/docs/user_guide/interaction/linking.html#userguide-interaction-linking
I assume that selecting points in the left plot, holding shift, and the select some points in the right plot should lead to the union selection in the source. It looks like the individual plots does union within itself ?

### PR DESCRIPTION
**Pull request by clairetang6**
This PR fixes the issue where a selection made on a plot containing multiple renderers that share the same data source only works with the last renderer added to the plot. The problem is that the selection manager writes over its data sources `.selected` property on every select call, but the select tool can make multiple select calls per selection event because the tools loop through the renderers and call select for each renderer. 

This PR fixes the issue by changing the `renderer_view` parameter for the select method on the selection manager to `renderer_views` so that multiple renderers (which are associated with that data source/selection manager) can be passed in together for one selection event. So then, each select tool groups its renderers by the data sources that they share and calls select for each group of renderers. 

Also, I removed the selection manager's `selectors` dict (which mapped renderers to selectors) in favor of a single `selector` because:

1. the `selected` rows of a column data source should be shared across all glyph renderers that use it.
2. each cds is associated with a selection manager in a one-to-one way.

This PR additionally fixes #5044, so that you can now shift-select to append to the selection across different plots when they have renderers that share the same data source.

- [x] issues: fixes #2136, #5044
- [x] tests added / passed

selection

```python
from bokeh.io import output_file, show
from bokeh.layouts import gridplot
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure

output_file("selection_multiple_renderers.html")

x = list(range(-20, 21))
y0 = [abs(xx) for xx in x]
y1 = [xx**2 for xx in x]
y2 = range(len(x))

source = ColumnDataSource(data=dict(x=x, y0=y0, y1=y1, y2=y2))

TOOLS = "box_select,lasso_select,help"

left = figure(tools=TOOLS, plot_width=300, plot_height=300, title=None)
left.circle('x', 'y0', source=source)
left.circle('x', 'y2', source=source, color='red')

right = figure(tools=TOOLS, plot_width=300, plot_height=300, title=None)
right.circle('x', 'y1', source=source)

p = gridplot([[left, right]])

show(p)
```

### CODE REVIEW

File: bokehjs/src/coffee/core/hittest.coffee [hunk 1]
```diff
@@ -19,6 +19,19 @@ export point_in_poly = (x, y, px, py) ->
 
 nullreturner = () -> null  # stub function shared by all hittests by default
 
+merge = (obj1, obj2) ->
```
Comment (mattpap) [inline bokehjs/src/coffee/core/hittest.coffee]: @clairetang6, can you move this to `core/util/object.ts` and add tests?

Comment (mattpap) [inline bokehjs/src/coffee/core/hittest.coffee]: The signature should be `function merge(obj1: object, obj2: object): object`.

Comment (clairetang6) [inline bokehjs/src/coffee/core/hittest.coffee]: Ok, will do.

**Comment by bryevdv** [comment]
@clairetang6 FYI a merge/rebase on master will fix up the current test failures

File: bokehjs/src/coffee/core/util/object.ts [hunk 2]
```diff
@@ -27,6 +29,21 @@ export function clone<T>(obj: T): T {
   return extend({}, obj)
 }
 
+export function merge<T>(obj1: {[key: string] : Array<T>}, obj2: {[key: string]: Array<T>}): object {
```
Comment (clairetang6) [inline bokehjs/src/coffee/core/util/object.ts]: @mattpap I tried the type signature with obj1 and obj2 being objects, but there was some sort of warning. This seems to have fixed it, but I'm not certain it's the right thing to do. Also, I don't know if there is common notion of what "merge" should mean for objects. I guess this function has to take objects with arrays as values and unions those array values for two objects.

Comment (bryevdv) [inline bokehjs/src/coffee/core/util/object.ts]: > Also, I don't know if there is common notion of what "merge" should mean for objects.

I was actually going to suggest it *not* go in `util/objects` for this reason (tho I think tests for it are good)

File: bokehjs/src/coffee/core/util/object.ts [hunk 3]
```diff
@@ -27,6 +29,21 @@ export function clone<T>(obj: T): T {
...
+  let result: {[key: string]: Array<T>} = Object.create(null);
+
+  let keys = concat([Object.keys(obj1), Object.keys(obj2)])
+
+  for (let i = 0; i < keys.length; i++) {
+    let key = keys[i]
```
Comment (mattpap) [inline bokehjs/src/coffee/core/util/object.ts]: `for (const key of keys)`

File: bokehjs/src/coffee/core/util/object.ts [hunk 4]
```diff
@@ -27,6 +29,21 @@ export function clone<T>(obj: T): T {
   return extend({}, obj)
 }
 
+export function merge<T>(obj1: {[key: string] : Array<T>}, obj2: {[key: string]: Array<T>}): object {
+  let result: {[key: string]: Array<T>} = Object.create(null);
```
Comment (mattpap) [inline bokehjs/src/coffee/core/util/object.ts]: `const`. The same elsewhere.

File: bokehjs/src/coffee/core/util/object.ts (hunk 2 above)
Comment (mattpap) [inline bokehjs/src/coffee/core/util/object.ts]: @clairetang6, I was wrong with the original signature. We define `merge()` for specific class of objects, so the definition here is correct. Though the return type could be more specific (the same as `obj1` and `obj2`).

File: bokehjs/src/coffee/core/util/object.ts [hunk 5]
```diff
@@ -27,6 +29,21 @@ export function clone<T>(obj: T): T {
...
+
+  let keys = concat([Object.keys(obj1), Object.keys(obj2)])
+
+  for (let i = 0; i < keys.length; i++) {
+    let key = keys[i]
+    let arr1: Array<T> = obj1.hasOwnProperty(key) ? obj1[key] : []
+    let arr2: Array<T> = obj2.hasOwnProperty(key) ? obj2[key] : []
```
Comment (mattpap) [inline bokehjs/src/coffee/core/util/object.ts]: Those type annotations shouldn't be necessary.

**Comment by bryevdv** [comment]
LGTM

File: bokehjs/src/coffee/models/tools/gestures/select_tool.coffee [hunk 6]
```diff
@@ -80,3 +80,12 @@ export class SelectTool extends GestureTool
...
     computed_renderers: () -> @_get_computed('computed_renderers')
   }
+
+  _computed_renderers_by_data_source: () ->
+    renderers_by_source = {}
+    for r in @computed_renderers
+      if !(r.data_source of renderers_by_source)
+        renderers_by_source[r.data_source] = [r]
+      else
+        renderers_by_source[r.data_source] = renderers_by_source[r.data_source].concat([r])
```
Comment (mattpap) [inline bokehjs/src/coffee/models/tools/gestures/select_tool.coffee]: Plain JS object can have only string keys. This should use `r.data_source.id`.

Comment (mattpap) [inline bokehjs/src/coffee/models/tools/gestures/select_tool.coffee]: Later on we should switch from using plain objects as dictionaries to `Map` or `WeakMap`.

**Comment by clairetang6** [comment]
@mattpap, yeah I had noticed that I couldn't get the selection manager straight from the data source key since it had been converted to a string. I'm assuming with the Map that then the key would be the actual object? Then, you could get the selection manager from the ds without having to go through a renderer (just ds.sm instead of r's[0].ds.sm).

**Comment by bryevdv** [comment]
Failing the file size limit test 
```
  1) js/bokeh.min.js file size should remain below limit:
     AssertionError: expected 666329 to be below 665600
      at Context.<anonymous> (test/size.coffee:22:37)
``` 

Will have to increase limit or find some savings somewhere.

**Comment by canavandl** [comment]
For some reason, the build size on my machine is (fractionally) lower on my machine than on Travis. So the tests will pass locally and not on CI.

**Comment by bryevdv** [comment]
@canavandl @clairetang6 I will fix the merge conflict and update the size limits this afternoon, then merge for a dev build.

**Comment by mattpap** [comment]
> the size limits this afternoon

The size limit was bumped to 700 kB in PR #6233, so it should be sufficient.

**Comment by mattpap** [comment]
```
<<<<<<< clairetang6/2136_selection_multiple_renderers
      source.trigger('select')

      for renderer_view in renderer_views
        source.trigger('select-' + renderer_view.model.id)
=======
      source.select.emit(renderer_view.model.id)
>>>>>>> master
```
I'm not sure of resolution of this. If we need those ids, then this would work:
```
source.select.emit([ view.model.id for view in renderer_views ])
```

**Comment by bryevdv** [comment]
@mattpap why would it not be
```
for renderer_view in renderer_views
 source.select.emit(renderer_view.model.id)
```
?

**Comment by mattpap** [comment]
> why would it not be

Because you are emitting the same signal multiple times, just with different argument. Then each slot connected to this signal will trigger multiple times. Currently there is only one slot connected in glyph_renderer (`request_render`, so you wouldn't even notice that something is wrong due to throttling). Given that no one currently uses those ids, I would just not emit them.

**Comment by clairetang6** [comment]
In the selection manager before this PR and before the change to signaling, there were two calls to source.trigger:

```python
source.trigger('select')
source.trigger('select-' + renderer_view.model.id
```

I don't think the second call with the id was listened to by anything in bokehjs, but I thought maybe it was there in case a user wanted to listen to a select event that was triggered by a specific renderer. Were users able to access the backbone events before? If they weren't then, could we just change that line to `source.select.emit()`?

**Comment by bryevdv** [comment]
@clairetang6 not easily and it was never demonstrated or advertised as a supported thing to do. I don't think it needs to be replaced.

**Comment by bryevdv** [comment]
@clairetang6 I only just saw that you removed `ready` do you want this reverted? What was left to do?

**Comment by clairetang6** [comment]
Hi @bryevdv, no I only removed ready to fix the merge conflict. I think any other work related to this could be in a separate PR.

**Pull request merged**